File: day02.dart
import "dart:io";


Map<String, int> countChars(String string) {
  final Map<String, int> count = string
    .split('')
    .fold({}, (acc, letter) {
      acc[letter] = (acc[letter] ?? 0) + 1;
      return acc;
    });
  return count;
}

int pt1(Iterable<String> ids) {
  var double = 0;
  var triple = 0;

  for (var id in ids) {
    final charCount = countChars(id);
    if (charCount.values.contains(2)) {
      double++;
    }
    if (charCount.values.contains(3)) {
      triple++;
    }
  }
  return double * triple;
}

List<int> stringDiffs(String s1, s2) {
  List<int> diffs = [];
  for (var i=0; i < s1.length; i++) {
    var char1 = s1[i];
    var char2 = s2[i];
    if (char1 != char2) {
      diffs.add(i);
    }
  }
  return diffs;
}

String pt2(Iterable<String> ids) {
  for (var i=0; i < ids.length; i++) {
    var id = ids.elementAt(i);
    var idsToCompare = ids.toList().sublist(i + 1);
    for (var otherId in idsToCompare) {
      var diffs = stringDiffs(id, otherId);
      if (diffs.length == 1) {
        var diffIndex = diffs[0];
        return id.substring(0, diffIndex) + id.substring(diffIndex + 1);
      }
    }
  }
  return "Uh oh, something went wrong";
}

main() async {
  final ids = await File('day02-input.txt').readAsLines();
  // pt 1
  print(pt1(ids));
  // pt 2
  print(pt2(ids));
}

File: day02.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func countChars(text string) map[string]int {
	count := map[string]int{}
	for _, char := range text {
		count[string(char)]++
	}
	return count
}

func pt1(ids []string) int {
	var double, triple int
	checkAgainst := map[int]*int{
		2: &double,
		3: &triple,
	}
	for _, id := range ids {
		charCount := countChars(id)
		for qty, pointer := range checkAgainst {
			for _, count := range charCount {
				if count == qty {
					*pointer++
					break
				}
			}
		}
	}
	return double * triple
}

func stringDiffs(s1, s2 string) []int {
	diffs := []int{}
	for i, char := range s1 {
		if string(char) != string(s2[i]) {
			diffs = append(diffs, i)
		}
	}
	return diffs
}

func pt2(ids []string) string {
	for i, id := range ids {
		idsToCompare := ids[i+1:]
		for _, otherID := range idsToCompare {
			diffs := stringDiffs(id, otherID)
			if len(diffs) == 1 {
				diffIndex := diffs[0]
				return id[:diffIndex] + id[diffIndex+1:]
			}
		}
	}
	return "Uh oh, something went wrong"
}

func main() {
	file, err := os.Open("day02-input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var ids []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		ids = append(ids, scanner.Text())
	}
	// pt 1
	fmt.Println(pt1(ids))
	// pt 2
	fmt.Println(pt2(ids))
}
